<a href="https://colab.research.google.com/github/Shanis185/favorita-sales-forecasting/blob/main/Time%20Series%20Forecasting.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**TIME SERIES DATA FORECASTING FAVOURITA DATASET**


In [30]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Importing necessary libraries


In [40]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import datetime
import plotly.express as px
import warnings
warnings.filterwarnings('ignore')

Getting Familarized with the Datasets

In [29]:
#function to display the details of datasets
def details(dataset,datasetname):
  print(f"dimension of {datasetname} {dataset.shape}")
  print("="*50)
  print()
  print(dataset.head())
  print("="*50)
  print()
  print(dataset.info())
  print("="*50)
  print()
  print(dataset.describe())
  print("="*50)
  print("Missing Values")
  print(dataset.isnull().sum())
  print("="*50)
  print()



Detail of oil.csv

In [32]:
#importing oil dataset from drive and displaying the details
oil_df=pd.read_csv("/content/drive/MyDrive/data/oil.csv")
details(oil_df,"oil.csv")


dimension of oil.csv (1218, 2)

         date  dcoilwtico
0  2013-01-01         NaN
1  2013-01-02   93.139999
2  2013-01-03   92.970001
3  2013-01-04   93.120003
4  2013-01-07   93.199997

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1218 entries, 0 to 1217
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        1218 non-null   object 
 1   dcoilwtico  1175 non-null   float64
dtypes: float64(1), object(1)
memory usage: 19.2+ KB
None

        dcoilwtico
count  1175.000000
mean     67.714366
std      25.630476
min      26.190001
25%      46.405001
50%      53.189999
75%      95.660000
max     110.620003
Missing Values
date           0
dcoilwtico    43
dtype: int64



Details of holiday events.csv

In [33]:
holiday_df=pd.read_csv("/content/drive/MyDrive/data/holidays_events.csv")
details(holiday_df,"holiday_events.csv")


dimension of holiday_events.csv (350, 6)

         date     type    locale locale_name                    description  \
0  2012-03-02  Holiday     Local       Manta             Fundacion de Manta   
1  2012-04-01  Holiday  Regional    Cotopaxi  Provincializacion de Cotopaxi   
2  2012-04-12  Holiday     Local      Cuenca            Fundacion de Cuenca   
3  2012-04-14  Holiday     Local    Libertad      Cantonizacion de Libertad   
4  2012-04-21  Holiday     Local    Riobamba      Cantonizacion de Riobamba   

   transferred  
0        False  
1        False  
2        False  
3        False  
4        False  

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   date         350 non-null    object
 1   type         350 non-null    object
 2   locale       350 non-null    object
 3   locale_name  350 non-null    object
 4   description  350 non-null

Details of stores.csv


In [34]:
stores_df=pd.read_csv("/content/drive/MyDrive/data/stores.csv")
details(stores_df,"stores.csv")


dimension of stores.csv (54, 5)

   store_nbr           city                           state type  cluster
0          1          Quito                       Pichincha    D       13
1          2          Quito                       Pichincha    D       13
2          3          Quito                       Pichincha    D        8
3          4          Quito                       Pichincha    D        9
4          5  Santo Domingo  Santo Domingo de los Tsachilas    D        4

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54 entries, 0 to 53
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   store_nbr  54 non-null     int64 
 1   city       54 non-null     object
 2   state      54 non-null     object
 3   type       54 non-null     object
 4   cluster    54 non-null     int64 
dtypes: int64(2), object(3)
memory usage: 2.2+ KB
None

       store_nbr    cluster
count  54.000000  54.000000
mean   27.500000   8.481481
std    

Details of transction.csv


In [35]:
transactions_df=pd.read_csv("/content/drive/MyDrive/data/transactions.csv")
details(transactions_df,"transactions.csv")


dimension of transactions.csv (83488, 3)

         date  store_nbr  transactions
0  2013-01-01         25           770
1  2013-01-02          1          2111
2  2013-01-02          2          2358
3  2013-01-02          3          3487
4  2013-01-02          4          1922

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 83488 entries, 0 to 83487
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          83488 non-null  object
 1   store_nbr     83488 non-null  int64 
 2   transactions  83488 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 1.9+ MB
None

          store_nbr  transactions
count  83488.000000  83488.000000
mean      26.939237   1694.602158
std       15.608204    963.286644
min        1.000000      5.000000
25%       13.000000   1046.000000
50%       27.000000   1393.000000
75%       40.000000   2079.000000
max       54.000000   8359.000000
Missing Values
date            0
store_nbr 

Details of train.csv

In [36]:
train_df=pd.read_csv("/content/drive/MyDrive/data/train.csv")
details(train_df,"train.csv")


dimension of train.csv (3000888, 6)

   id        date  store_nbr      family  sales  onpromotion
0   0  2013-01-01          1  AUTOMOTIVE    0.0            0
1   1  2013-01-01          1   BABY CARE    0.0            0
2   2  2013-01-01          1      BEAUTY    0.0            0
3   3  2013-01-01          1   BEVERAGES    0.0            0
4   4  2013-01-01          1       BOOKS    0.0            0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000888 entries, 0 to 3000887
Data columns (total 6 columns):
 #   Column       Dtype  
---  ------       -----  
 0   id           int64  
 1   date         object 
 2   store_nbr    int64  
 3   family       object 
 4   sales        float64
 5   onpromotion  int64  
dtypes: float64(1), int64(3), object(2)
memory usage: 137.4+ MB
None

                 id     store_nbr         sales   onpromotion
count  3.000888e+06  3.000888e+06  3.000888e+06  3.000888e+06
mean   1.500444e+06  2.750000e+01  3.577757e+02  2.602770e+00
std    8.662819e+05 

Details of the test.csv

In [37]:
test_df=pd.read_csv("/content/drive/MyDrive/data/test.csv")
details(test_df,"test.csv")


dimension of test.csv (28512, 5)

        id        date  store_nbr      family  onpromotion
0  3000888  2017-08-16          1  AUTOMOTIVE            0
1  3000889  2017-08-16          1   BABY CARE            0
2  3000890  2017-08-16          1      BEAUTY            2
3  3000891  2017-08-16          1   BEVERAGES           20
4  3000892  2017-08-16          1       BOOKS            0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28512 entries, 0 to 28511
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   id           28512 non-null  int64 
 1   date         28512 non-null  object
 2   store_nbr    28512 non-null  int64 
 3   family       28512 non-null  object
 4   onpromotion  28512 non-null  int64 
dtypes: int64(3), object(2)
memory usage: 1.1+ MB
None

                 id     store_nbr   onpromotion
count  2.851200e+04  28512.000000  28512.000000
mean   3.015144e+06     27.500000      6.965383
std    8.230850e

By looking upon the details we came to know that the datatype of date is object for exploitation the datatype should be changed to datetime


There are 43 missing values in oil price in oil.csv

Transforming Object to Datetime

In [59]:
def datetimetransform(dataset):
  dataset["date"]=pd.to_datetime(dataset["date"])
datetimetransform(oil_df)
datetimetransform(holiday_df)
datetimetransform(transactions_df)
datetimetransform(train_df)
datetimetransform(test_df)

Handling missing values

In [60]:
#visualize the oil.csv to see how large the gaps are
fig=px.line(oil_df,x='date',y="dcoilwtico")
fig.update_layout(title="Trend of oil prices",xaxis_title="Date",yaxis_title='oil price')
fig.update_xaxes(rangeslider_visible=True)
fig.show()

Backward fill is suitable as there are irregular gaps in  the plot an its the most suitable way as the future values are the best estimator for filling the gap

In [42]:
oil_df['dcoilwtico']=oil_df['dcoilwtico'].fillna(method='bfill')

Cross checking for missing values in oil.csv

In [44]:
details(oil_df,"oil.csv")

dimension of oil.csv (1218, 2)

        date  dcoilwtico
0 2013-01-01   93.139999
1 2013-01-02   93.139999
2 2013-01-03   92.970001
3 2013-01-04   93.120003
4 2013-01-07   93.199997

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1218 entries, 0 to 1217
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   date        1218 non-null   datetime64[ns]
 1   dcoilwtico  1218 non-null   float64       
dtypes: datetime64[ns](1), float64(1)
memory usage: 19.2 KB
None

                      date   dcoilwtico
count                 1218  1218.000000
mean   2015-05-02 12:00:00    67.673325
min    2013-01-01 00:00:00    26.190001
25%    2014-03-03 06:00:00    46.422500
50%    2015-05-02 12:00:00    53.184999
75%    2016-06-30 18:00:00    95.595001
max    2017-08-31 00:00:00   110.620003
std                    NaN    25.637345
Missing Values
date          0
dcoilwtico    0
dtype: int64



Check the continouity of the Date in the test.csv

In [48]:
min_date = train_df['date'].min()
max_date = train_df['date'].max()
expected_dates = pd.date_range(start=min_date, end=max_date)

missing_dates = expected_dates[~expected_dates.isin(train_df['date'])]

if len(missing_dates) == 0:
    print("The train dataset is complete. It includes all the required dates.")
else:
    print("The train dataset is incomplete. The following dates are missing:")
    print(missing_dates)


The train dataset is incomplete. The following dates are missing:
DatetimeIndex(['2013-12-25', '2014-12-25', '2015-12-25', '2016-12-25'], dtype='datetime64[ns]', freq=None)


handle the missing dates

In [49]:
# Complete the missing dates in the train dataset
# Create an index of the missing dates as a DatetimeIndex object
missing_dates = pd.Index(['2013-12-25', '2014-12-25', '2015-12-25', '2016-12-25'], dtype='datetime64[ns]')

# Create a DataFrame with the missing dates, using the 'date' column
missing_data = pd.DataFrame({'date': missing_dates})

# Concatenate the original train dataset and the missing data DataFrame
# ignore_index=True ensures a new index is assigned to the resulting DataFrame
train_df = pd.concat([train_df, missing_data], ignore_index=True)

# Sort the DataFrame based on the 'date' column in ascending order
train_df.sort_values('date', inplace=True)

confirming the dataset


In [50]:
min_date = train_df['date'].min()
max_date = train_df['date'].max()
expected_dates = pd.date_range(start=min_date, end=max_date)

missing_dates = expected_dates[~expected_dates.isin(train_df['date'])]

if len(missing_dates) == 0:
    print("The train dataset is complete. It includes all the required dates.")
else:
    print("The train dataset is incomplete. The following dates are missing:")
    print(missing_dates)


The train dataset is complete. It includes all the required dates.


Merge the oil.csv,holiday_events.csv,tranaction.csv to the train.csv using inner join on date coloumn

merge store.csv using store_nbr column

In [52]:
# Merging the common columns ('store_nbr' and 'date') in the datasets using the inner merge() function
# Merge train_df with stores_df based on 'store_nbr' column
merged_df1 = train_df.merge(stores_df, on='store_nbr', how='inner')

# Merge merged_df1 with transactions_df based on 'date' and 'store_nbr' columns
merged_df2 = merged_df1.merge(transactions_df, on=['date', 'store_nbr'], how='inner')

# Merge merged_df2 with holidays_events_df based on 'date' column
merged_df3 = merged_df2.merge(holiday_df, on='date', how='inner')

# Merge merged_df3 with oil_df based on 'date' column
merged_df = merged_df3.merge(oil_df, on='date', how='inner')

# View the first five rows of the merged dataset
merged_df.head()

id       date  store_nbr               family  sales  onpromotion  \
0  581.0 2013-01-01       25.0      LAWN AND GARDEN    2.0          0.0   
1  580.0 2013-01-01       25.0           LADIESWEAR    0.0          0.0   
2  579.0 2013-01-01       25.0            HOME CARE    0.0          0.0   
3  578.0 2013-01-01       25.0      HOME APPLIANCES    0.0          0.0   
4  577.0 2013-01-01       25.0  HOME AND KITCHEN II    0.0          0.0   

      city        state type_x  cluster  transactions   type_y    locale  \
0  Salinas  Santa Elena      D        1           770  Holiday  National   
1  Salinas  Santa Elena      D        1           770  Holiday  National   
2  Salinas  Santa Elena      D        1           770  Holiday  National   
3  Salinas  Santa Elena      D        1           770  Holiday  National   
4  Salinas  Santa Elena      D        1           770  Holiday  National   

  locale_name         description  transferred  dcoilwtico  
0     Ecuador  Primer dia del ano        False   93.139999  
1     Ecuador  Primer dia del ano        False   93.139999  
2     Ecuador  Primer dia del ano        False   93.139999  
3     Ecuador  Primer dia del ano        False   93.139999  
4     Ecuador  Primer dia del ano        False   93.139999

The use of an inner merge in this time series forecasting project for corporation Favorita helps to ensure data consistency, avoid missing values, and focus on the relevant data for accurate predictions.

An inner merge type retains only the rows with matching values in the specified columns. In the context of time series forecasting, it allows us to merge datasets based on a common time index or timestamp. By performing an inner merge, we ensure that only the rows with corresponding timestamps in both datasets are included in the merged result. This is important for time series forecasting because you want to align the data from different sources based on their timestamps to build a consistent and accurate forecasting model.

With an inner merge, you eliminate any non-matching timestamps, which may not be useful for forecasting and could introduce inconsistencies in the data. By focusing on the intersection of the datasets, we can create a merged dataset that contains the necessary information for accurate time series forecasting.

checking the details of merged dataset


In [53]:
details(merged_df,"merged_df")


dimension of merged_df (322047, 17)

      id       date  store_nbr               family  sales  onpromotion  \
0  581.0 2013-01-01       25.0      LAWN AND GARDEN    2.0          0.0   
1  580.0 2013-01-01       25.0           LADIESWEAR    0.0          0.0   
2  579.0 2013-01-01       25.0            HOME CARE    0.0          0.0   
3  578.0 2013-01-01       25.0      HOME APPLIANCES    0.0          0.0   
4  577.0 2013-01-01       25.0  HOME AND KITCHEN II    0.0          0.0   

      city        state type_x  cluster  transactions   type_y    locale  \
0  Salinas  Santa Elena      D        1           770  Holiday  National   
1  Salinas  Santa Elena      D        1           770  Holiday  National   
2  Salinas  Santa Elena      D        1           770  Holiday  National   
3  Salinas  Santa Elena      D        1           770  Holiday  National   
4  Salinas  Santa Elena      D        1           770  Holiday  National   

  locale_name         description  transferred  dcoilwt

The merged dataset after merging the train dataset with additional datasets contains 322,047 rows and 16 columns. Two columns have been renamed as a result of the merging, type_x and type_y.

checking for the details of the newly created columns


In [54]:
# Check the unique values of the two unknown columns
print("Unique values of 'type_x':")
print(merged_df['type_x'].unique())
print()
print("Unique values of 'type_y':")
print(merged_df['type_y'].unique())

Unique values of 'type_x':
['D' 'E' 'C' 'A' 'B']

Unique values of 'type_y':
['Holiday' 'Additional' 'Transfer' 'Event' 'Bridge']


Unique values of 'type_x':
['D' 'E' 'C' 'A' 'B']

Unique values of 'type_y':

['Holiday' 'Additional' 'Transfer' 'Event' 'Bridge']

The merged dataset consists of 322,047 non-null observations. Two columns have been renamed as a result of the merging, type_x and type_y.

The type_x column represents the store type.

The type_y column represents the holiday type.

In [55]:
merged_df = merged_df.rename(columns={"type_x": "store_type", "type_y": "holiday_type"})
merged_df.head()

id       date  store_nbr               family  sales  onpromotion  \
0  581.0 2013-01-01       25.0      LAWN AND GARDEN    2.0          0.0   
1  580.0 2013-01-01       25.0           LADIESWEAR    0.0          0.0   
2  579.0 2013-01-01       25.0            HOME CARE    0.0          0.0   
3  578.0 2013-01-01       25.0      HOME APPLIANCES    0.0          0.0   
4  577.0 2013-01-01       25.0  HOME AND KITCHEN II    0.0          0.0   

      city        state store_type  cluster  transactions holiday_type  \
0  Salinas  Santa Elena          D        1           770      Holiday   
1  Salinas  Santa Elena          D        1           770      Holiday   
2  Salinas  Santa Elena          D        1           770      Holiday   
3  Salinas  Santa Elena          D        1           770      Holiday   
4  Salinas  Santa Elena          D        1           770      Holiday   

     locale locale_name         description  transferred  dcoilwtico  
0  National     Ecuador  Primer dia del ano        False   93.139999  
1  National     Ecuador  Primer dia del ano        False   93.139999  
2  National     Ecuador  Primer dia del ano        False   93.139999  
3  National     Ecuador  Primer dia del ano        False   93.139999  
4  National     Ecuador  Primer dia del ano        False   93.139999

checking details of merged dataset


In [56]:
details(merged_df,"merged_df")

dimension of merged_df (322047, 17)

      id       date  store_nbr               family  sales  onpromotion  \
0  581.0 2013-01-01       25.0      LAWN AND GARDEN    2.0          0.0   
1  580.0 2013-01-01       25.0           LADIESWEAR    0.0          0.0   
2  579.0 2013-01-01       25.0            HOME CARE    0.0          0.0   
3  578.0 2013-01-01       25.0      HOME APPLIANCES    0.0          0.0   
4  577.0 2013-01-01       25.0  HOME AND KITCHEN II    0.0          0.0   

      city        state store_type  cluster  transactions holiday_type  \
0  Salinas  Santa Elena          D        1           770      Holiday   
1  Salinas  Santa Elena          D        1           770      Holiday   
2  Salinas  Santa Elena          D        1           770      Holiday   
3  Salinas  Santa Elena          D        1           770      Holiday   
4  Salinas  Santa Elena          D        1           770      Holiday   

     locale locale_name         description  transferred  dcoilwtic

**SAVING THE MERGED DATASET**

In [57]:
merged_df.to_csv('/content/drive/MyDrive/data/Visualization_Data.csv', index=False)
merged_df.head()

id       date  store_nbr               family  sales  onpromotion  \
0  581.0 2013-01-01       25.0      LAWN AND GARDEN    2.0          0.0   
1  580.0 2013-01-01       25.0           LADIESWEAR    0.0          0.0   
2  579.0 2013-01-01       25.0            HOME CARE    0.0          0.0   
3  578.0 2013-01-01       25.0      HOME APPLIANCES    0.0          0.0   
4  577.0 2013-01-01       25.0  HOME AND KITCHEN II    0.0          0.0   

      city        state store_type  cluster  transactions holiday_type  \
0  Salinas  Santa Elena          D        1           770      Holiday   
1  Salinas  Santa Elena          D        1           770      Holiday   
2  Salinas  Santa Elena          D        1           770      Holiday   
3  Salinas  Santa Elena          D        1           770      Holiday   
4  Salinas  Santa Elena          D        1           770      Holiday   

     locale locale_name         description  transferred  dcoilwtico  
0  National     Ecuador  Primer dia del ano        False   93.139999  
1  National     Ecuador  Primer dia del ano        False   93.139999  
2  National     Ecuador  Primer dia del ano        False   93.139999  
3  National     Ecuador  Primer dia del ano        False   93.139999  
4  National     Ecuador  Primer dia del ano        False   93.139999